Add dynamically updating routes example (#1471)

tuleism opened the pull request:
Ref #1471.

raboof commented on docs/src/test/scala/docs/http/scaladsl/HttpServerExampleSpec.scala:
```diff
@@ -699,5 +699,49 @@ class HttpServerExampleSpec extends WordSpec with Matchers
+
+    // dynamic routing based on current state
+    val dynamicRoute: Route = ctx => {
+      val routes = state.map { kv =>
```
Perhaps use `case (path, response)` notation here?

raboof commented on docs/src/test/scala/docs/http/scaladsl/HttpServerExampleSpec.scala:
```diff
@@ -699,5 +699,49 @@ class HttpServerExampleSpec extends WordSpec with Matchers
+      pathSingleSlash {
+        entity(as[MockDefinition]) { mock =>
+          val mapping = mock.requests.zip(mock.responses).toMap
+          state.put(mock.path, mapping)
```
While `TrieMap` is indeed thread-safe, it might be a better example to replace the `Map`?

jlprat commented:
OK TO TEST

tuleism commented:
@raboof 
- I've addressed your comments on the Scala code.
- I've added a Java version of the example. This is mostly copied from other examples as I'm not familiar with Java. Please see if it makes sense.
- I also change the type to `JsValue` to be more general than Json object, and change the documentation to be more focused on the route definition.

raboof commented on docs/src/test/java/docs/http/javadsl/HttpServerDynamicRoutingExampleTest.java:
```diff
@@ -0,0 +1,111 @@
+
+    final Flow<HttpRequest, HttpResponse, NotUsed> routeFlow = app.createRoute().flow(system, materializer);
+    final CompletionStage<ServerBinding> binding = http.bindAndHandle(routeFlow,
+      ConnectHttp.toHost("localhost", 8080), materializer);
```
I'd say it should be sufficient to just end the `main` method after `http.bindAndHandle(routeFlow,
ConnectHttp.toHost("localhost", 8080), materializer);` and leave it up to the user to kill the application, though to make that convenient we should add sbt-revolver to our build

raboof commented on docs/src/test/scala/docs/http/scaladsl/HttpServerExampleSpec.scala:
```diff
@@ -699,5 +699,50 @@ class HttpServerExampleSpec extends WordSpec with Matchers
+      pathSingleSlash {
+        entity(as[MockDefinition]) { mock =>
+          val mapping = mock.requests.zip(mock.responses).toMap
+          state += mock.path -> mapping
```
perhaps even `state = state + (mock.path -> mapping)`?

(your example is indeed equally safe, but that might not be obvious to people newer to Scala)

tuleism commented:
@raboof 
Both new comments are addressed.

ktoso commented on docs/src/test/scala/docs/http/scaladsl/HttpServerExampleSpec.scala:
```diff
@@ -699,5 +699,50 @@ class HttpServerExampleSpec extends WordSpec with Matchers
+    case class MockDefinition(path: String, requests: Seq[JsValue], responses: Seq[JsValue])
+    implicit val format = jsonFormat3(MockDefinition)
+
+    var state = Map.empty[String, Map[JsValue, JsValue]]
```
hmm mutating this Map from potentially multiple threads -- since one handler flow is materialized for every connection, so there's many at once. Should this be made volatile for the example?

tuleism replied:
Yes, done!

johanandren commented on docs/src/test/java/docs/http/javadsl/HttpServerDynamicRoutingExampleTest.java:
```diff
@@ -0,0 +1,102 @@
+  }
+
+  //#dynamic-routing-example
+  private Map<String, Map<JsonNode, JsonNode>> state = new HashMap<>();
```
This also needs to be made thread safe - `final private ... = new ConcurrentHashMap<>();`

johanandren commented on docs/src/main/paradox/routing-dsl/index.md:
```diff
@@ -69,6 +69,36 @@ the Routing DSL will look like:
 
 @@@
 
+## Dynamic Routing Example
```
Would be good with an introductory paragraph saying that as the routes are evaluated for each request it is possible to make it change at runtime. However every access may happen on a separate thread so any shared mutable state must be thread safe.

tuleism commented:
@johanandren I've fixed for the Java example and the introductory paragraph.

johanandren commented on docs/src/main/paradox/routing-dsl/index.md:
```diff
@@ -69,6 +69,38 @@ the Routing DSL will look like:
+## Dynamic Routing Example
+
+As the routes are evaluated for each request, it is possible to make changes at runtime. Please note that every access
+may happen on a separated thread, so any shared mutable state must be thread safe.
```
Great

ktoso replied:
Good note, thanks
